File: VB/PersonalOrganizer/ContactContextDataModel/UnitOfWorkSource.vb
Imports System
Imports System.Linq
Imports System.Data
Imports System.Data.Entity
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports Common.Utils
Imports Common.DataModel
Imports Common.DataModel.EntityFramework
Imports PersonalOrganizer.Model
Imports DevExpress.Mvvm
Namespace ContactContextDataModel
    Public Module UnitOfWorkSource
        Friend Class DbUnitOfWorkFactory
            Implements IUnitOfWorkFactory
            Public Shared ReadOnly Instance As IUnitOfWorkFactory = New DbUnitOfWorkFactory()
            Private Sub New()
            End Sub
            Private Function CreateUnitOfWork() As IContactContextUnitOfWork Implements IUnitOfWorkFactory(Of IContactContextUnitOfWork).CreateUnitOfWork
                Return New ContactContextUnitOfWork(New ContactContext())
            End Function
        End Class
        Friend Class DesignUnitOfWorkFactory
            Implements IUnitOfWorkFactory
            Public Shared ReadOnly Instance As IUnitOfWorkFactory = New DesignUnitOfWorkFactory()
            Private ReadOnly _UnitOfWork As IContactContextUnitOfWork = New ContactContextDesignTimeUnitOfWork()
            Private Sub New()
            End Sub
            Private Function CreateUnitOfWork() As IContactContextUnitOfWork Implements IUnitOfWorkFactory(Of IContactContextUnitOfWork).CreateUnitOfWork
                Return _UnitOfWork
            End Function
        End Class
        Public Function GetUnitOfWorkFactory() As IUnitOfWorkFactory
            Return GetUnitOfWorkFactory(ViewModelBase.IsInDesignMode)
        End Function
        Public Function GetUnitOfWorkFactory(ByVal isInDesignTime As Boolean) As IUnitOfWorkFactory
            Return If(isInDesignTime, DesignUnitOfWorkFactory.Instance, DbUnitOfWorkFactory.Instance)
        End Function
        Public Function CreateUnitOfWork(Optional ByVal isInDesignTime As Boolean = False) As IContactContextUnitOfWork
            Return GetUnitOfWorkFactory(isInDesignTime).CreateUnitOfWork()
        End Function
    End Module
End Namespace

File: CS/PersonalOrganizer/ContactContextDataModel/UnitOfWorkSource.cs
using System;
using System.Linq;
using System.Data;
using System.Data.Entity;
using System.Linq.Expressions;
using System.Collections.Generic;
using PersonalOrganizer.Common.Utils;
using PersonalOrganizer.Common.DataModel;
using PersonalOrganizer.Common.DataModel.EntityFramework;
using PersonalOrganizer.Model;
using DevExpress.Mvvm;

namespace PersonalOrganizer.ContactContextDataModel {
    public static class UnitOfWorkSource {
        #region inner classes
        class DbUnitOfWorkFactory : IUnitOfWorkFactory {
            public static readonly IUnitOfWorkFactory Instance = new DbUnitOfWorkFactory();
            DbUnitOfWorkFactory() { }
            IContactContextUnitOfWork IUnitOfWorkFactory<IContactContextUnitOfWork>.CreateUnitOfWork() {
                return new ContactContextUnitOfWork(new ContactContext());
            }
        }
        class DesignUnitOfWorkFactory : IUnitOfWorkFactory {
            public static readonly IUnitOfWorkFactory Instance = new DesignUnitOfWorkFactory();

            readonly IContactContextUnitOfWork UnitOfWork = new ContactContextDesignTimeUnitOfWork();
            DesignUnitOfWorkFactory() { }
            IContactContextUnitOfWork IUnitOfWorkFactory<IContactContextUnitOfWork>.CreateUnitOfWork() {
                return UnitOfWork;
            }
        }
        #endregion
        public static IUnitOfWorkFactory GetUnitOfWorkFactory() {
            return GetUnitOfWorkFactory(ViewModelBase.IsInDesignMode);
        }
        public static IUnitOfWorkFactory GetUnitOfWorkFactory(bool isInDesignTime) {
            return isInDesignTime ? DesignUnitOfWorkFactory.Instance : DbUnitOfWorkFactory.Instance;
        }
        public static IContactContextUnitOfWork CreateUnitOfWork(bool isInDesignTime = false) {
            return GetUnitOfWorkFactory(isInDesignTime).CreateUnitOfWork();
        }
    }
}